shelley/chain-and-ledger/executable-spec/test/Generator/Core/QuickCheck.hs

```diff
@@ -0,0 +1,141 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Generator.Core.QuickCheck
+  ( findPayKeyPair
+  , genBool
+  , genCoin
+  , genInteger
+  , genNatural
+  , genTxOut
+  , genUtxo0
+  , increasingProbabilityAt
+  , mkGenesisLedgerState
+  , traceKeyPairs
+  , traceVRFKeyPairs
+  , someKeyPairs
+  , pickStakeKey
+  , toAddr
+  , toCred)
+  where
+
+import           Cardano.Crypto.VRF (deriveVerKeyVRF, genKeyVRF)
+import           Control.Monad (replicateM)
+import           Crypto.Random (drgNewTest, withDRG)
+import           Data.Tuple (swap)
+import           Data.Word (Word64)
+
+import           Test.QuickCheck (Gen)
+import qualified Test.QuickCheck as QC
+
+import           Address (toAddr, toCred)
+import           Coin (Coin (..))
+import           Examples (mkKeyPair)
+import           Keys (pattern KeyPair, hashKey, vKey)
+import           LedgerState (pattern LedgerState, genesisCoins, genesisState)
+import           MockTypes (Addr, DPState, KeyPair, KeyPairs, LedgerEnv, SignKeyVRF, TxOut, UTxO,
+                     UTxOState, VKey, VerKeyVRF)
+import           Numeric.Natural (Natural)
+import           Tx (pattern TxOut)
+import           TxData (pattern AddrBase, pattern KeyHashObj)
+
+genBool :: Gen Bool
+genBool = QC.arbitraryBoundedRandom
+
+genInteger :: Integer -> Integer -> Gen Integer
+genInteger lower upper = QC.choose (lower, upper)
+
+-- | Generator for a natural number between 'lower' and 'upper'
+genNatural :: Natural -> Natural -> Gen Natural
+genNatural lower upper = fromInteger <$> QC.choose (lower', upper')
+ where
+  lower' = fromIntegral lower
+  upper' = fromIntegral upper
+
+mkKeyPairs :: Word64 -> (KeyPair, KeyPair)
```

Comment:
maybe add a comment here and / or rename it. I see the reason for the plural, but somehow would expect a list to be generated. (but that's really a very minor comment- and this is probably something you didn't write for this PR but copied, right?)

Reply:
Yeah, this is copied from the original (hedgehog version) module. I agree that “s” usually means a list, rather than two like-types pairs, and this could be renamed. But it would be an unrelated change.